bip-0331.mediawiki

```diff
@@ -0,0 +1,373 @@
+output(s) from a transaction they are unaware of.  Orphans are common for new nodes that have just
+completed Initial Block Download (IBD) and do not have an up-to-date mempool.
+
+Nodes may handle orphans by storing them in a cache and requesting any missing parent(s) by txid
+(prevouts specify txid, not wtxid). These parents may end up being orphans as well, if they also
+spend unconfirmed inputs that the node is unaware of. This method of handling orphans is problematic
+for two reasons: it requires nodes to allocate memory for unvalidated data received on the p2p
+network and txid-based relay between two wtxid-relay peers.
+
+This proposal makes orphan-fetching more efficient and no longer require txid-based relay.
+
+==Definitions==
+
+Given any two transactions Tx0 and Tx1 where Tx1 spends an output of Tx0, Tx0 is a '''parent''' of
+Tx1 and Tx1 is a '''child''' of Tx0.
+
+A transaction's '''ancestors''' include, recursively, its parents, the parents of its parents, etc.
+A transaction's '''descendants''' include, recursively, its children, the children of its children,
+etc. A transaction's parent is its ancestor, but an ancestor is not necessarily a parent.
+
+A '''package''' is an ordered list of transactions, representable by a connected Directed Acyclic
+Graph (a directed edge exists between a transaction that spends the output of another transaction).
+Typically, a package refers to unconfirmed transactions.
+
+An '''ancestor package''' consists of an unconfirmed transaction with all of its unconfirmed
+ancestors.
+
+==Specification==
+
+===Intended Protocol Flow===
+
+Package Relay consists of two phases: a package information round and a transaction data round.
+<ref>'''Why are package information and transaction data rounds both necessary?'''
+
+Several alternative designs were considered. One should measure alternative solutions based on the
+resources used to communicate (not necessarily trustworthy) information: We would like to minimize
+network bandwidth, avoid downloading a transaction more than once, avoid downloading transactions
+that are eventually rejected, and minimize storage allocated for not-yet-validated transactions.
+For example, consider the following scenarios (where the mempool minimum feerate is 3sat/vB and all
+transactions are the same size):
+
+* Package {A, B} where A pays 0 satoshis and B pays 8000 satoshis in fees.
+* Package {C, D} where C pays 0 satoshis and D pays 1200 satoshis in fees.
+* Package {E, F, G, H, J} that pays 4000, 8000, 0, 2000, and 4000 satoshis in fees, respectively.
+[[File:./bip-package-relay/tx_scenarios.png|800px]]
+
+'''No Package Information Round:''' Instead of having a package information round, just use the
+child's wtxid to refer to the package and always send the entire package together. This would make
+it very likely for honest nodes to redownload duplicate transactions.
+[[File:./bip-package-relay/no_package_info.png|900px]]
```

Comment:
Need to update paths to images now that bip number is assigned

Reply:
Doh, thanks!